### DISCUSSION TITLE
Conditional imputation: Two solutions

### DISCUSSION THREAD

**Discussion by angelrodriguez2020** [amices/mice#354] [Missing data methodology]
Dear fellows,

I'm using your excellent package MICE. 

I have two variables that I need to impute together: `caidas` is the number of falls (numeric), and `cq18b` is a variable of the care received by the person who fell; it is a factor with levels 

1. No fall 
2. Medical care received
3. No medical care needed

When imputing `caidas`, I want that if `caidas==0`, `cq18b==1` and if `caidas>0` `cq18b` gets either an imputed  2 or 3. 

This fragment of code limits the second imputation to 3, but I'd like that it imputes 2 or 3, depending on other predictors.

```r
post["cq18b"] <- "
                  imp[[j]][data$caidas[!r[,j]]==0,i] <- levels(tshauprf2$cq18b)[1]
                  imp[[j]][data$caidas[!r[,j]]>0,i] <- levels(tshauprf2$cq18b)[3]
                        "
```
Any idea?

Thank you for your time and attention,
Ángel

**Comment by stefvanbuuren** [comment]
Thanks. 

Conditional imputation is recurring issue (#43, #125, #152, #153, #193) that has no elegant solution in `mice`. Part of the difficulty is that there are so many use cases, so we often need some form of tailoring. In the example below I demo conditional imputation with `mice` using the `post`  processing option.

```r
library(mice, warn.conflicts = FALSE)

data <- data.frame(
  age = c(60, 54, 82, NA, 34, 58, 60, 58, 76, 40),
  fall = c(0, NA, 1, 0, 1, NA, 1, 1, 0, 1),
  aid = factor(c("-", NA, "Yes", "-", "No", NA, "Yes", NA, NA, "No"))
)

pred <- make.predictorMatrix(data)
pred["fall", "aid"] <- 0
pred["age", "aid"] <- 0

post <- make.post(data)
post["aid"] <- paste(sep = "; ",
                     "fall <- data$fall[!r[,j]] == 1",
                     "idx <- data$fall == 1",
                     "imp[[j]][!fall, i] <- levels(data$aid)[1]",
                     "imp[[j]][fall, i] <- mice.impute.pmm(y = data$aid[idx], ry = r[idx, j], x = model.matrix(aid ~ 0 + age, data[idx, , drop = FALSE]))")
                     
imp <- mice(data, m = 10, pred = pred, post = post, print = FALSE, seed = 1)
```

```r
> data
   age fall  aid
1   60    0    -
2   54   NA <NA>
3   82    1  Yes
4   NA    0    -
5   34    1   No
6   58   NA <NA>
7   60    1  Yes
8   58    1 <NA>
9   76    0 <NA>
10  40    1   No
```
There are several features that need commenting:

1. The `data` contain 10 rows. If `fall` is missing, then `aid` is also missing. The procedure does not deal with the case where `fall` is missing and `aid` is observed. If `fall` is zero, we will want to impute category `-`. If `fall` is 1, then we want to build the imputation model for `aid` on the subset of cases with `fall == 1`. If `fall` is missing, then we want to impute it. To maintain consistency among the imputes, I have placed the `aid` column adjacent to the `fall` column (alternatively, we may also tamper with the `visitSequence`).
2. We need to remove `aid` from the imputation models for `fall` and `age`. Otherwise the problem is multi-collinear. Although `mice` will detect and handle this, it is better to explicitly control the model.
3. The `post` statement will fit two imputation models for `aid`. If `fall = 0` then it imputes simply `-`, if `fall = 1` it imputes `Yes` or `No`.

```r
> complete(imp)
   age fall aid
1   60    0   -
2   54    0   -
3   82    1 Yes
4   34    0   -
5   34    1  No
6   58    1 Yes
7   60    1 Yes
8   58    1 Yes
9   76    0   -
10  40    1  No
> imp$imp
$age
   1  2  3  4  5  6  7  8  9 10
4 34 76 54 76 60 58 60 60 82 82

$fall
  1 2 3 4 5 6 7 8 9 10
2 0 1 1 1 1 1 1 1 0  1
6 1 1 0 1 1 1 0 1 1  1

$aid
    1   2   3   4   5   6   7   8   9  10
2   -  No  No  No Yes Yes Yes  No   -  No
6 Yes  No   - Yes  No Yes   - Yes Yes  No
8 Yes Yes Yes Yes  No Yes  No Yes Yes Yes
9   -   -   -   -   -   -   -   -   -   -
```

1. The first imputed dataset shows that the `Yes/No` imputes are only done for those columns with `fall` equal to 1, as intended. The remaining entries are imputed `-`.
2. Rows 2 and 6 have missing `fall`, which is then imputed as either `0` or `1`. Consequently, the `aid` column in the 10 imputed datasets can contain all three values: `Yes`, `No` or `-`, depending on how `fall` was imputed.
3. Row 8 had an observed fall, so imputes for `aid` are always `Yes` or `No`.
4. Row 9 has an observed no fall, so imputes for `aid` are always `-`.

I admit the procedure given above is somewhat complicated, but it does the job, and can be generalised to more difficult problems, for example, to three or more groups and for blocks of variables. Watch out for the case where one of the groups can become empty.

An alternative is to write your own `mice.impute.myimpute()` function. For example, copy `mice::mice.impute.norm()`, adapt it as you see fit, store it in your active work space under the name `mice.impute.myimpute()`, and call `mice` with the proper elements in `meth` set to `"myimpute"`.

Hope this is useful.

**Comment by stefvanbuuren** [comment]
And here's a solution based on a custom imputation function.

```r
library(mice, warn.conflicts = FALSE)

data <- data.frame(
  age = c(60, 54, 82, NA, 34, 58, 60, 58, 76, 40),
  fall = c(0, NA, 1, 0, 1, NA, 1, 1, 0, 1),
  aid = factor(c("-", NA, "Yes", "-", "No", NA, "Yes", NA, NA, "No"))
)

pred <- make.predictorMatrix(data)
pred["fall", "aid"] <- 0
pred["age", "aid"] <- 0

meth <- make.method(data)
meth$aid <- "aid"

mice.impute.aid <- function(y, ry, x, wy = NULL, ...) {
  if (is.null(wy)) wy <- !ry
  
  fall <- x[, "fall"][!ry] == 1
  idx <- x[, "fall"] == 1
  vec <- rep(NA, sum(wy))
  
  vec[!fall] <- 1
  vec[fall] <- as.integer(mice.impute.pmm(y[idx], ry = ry[idx], x = x[idx, "age", drop = FALSE], ...))
  
  levels(y)[vec]
}

imp <- mice(data, m = 10, maxit = 5, pred = pred, meth = meth, print = FALSE, seed = 1)
```

```r
> imp$imp
$age
   1  2  3  4  5  6  7  8  9 10
4 34 60 58 40 82 34 76 58 34 58

$fall
  1 2 3 4 5 6 7 8 9 10
2 0 0 0 1 1 1 0 1 0  1
6 1 1 1 1 1 1 1 1 1  0

$aid
    1   2   3   4   5   6   7   8  9  10
2   -   -   - Yes  No  No   -  No  - Yes
6 Yes Yes Yes Yes Yes Yes Yes Yes No   -
8 Yes  No  No Yes  No  No  No Yes No Yes
9   -   -   -   -   -   -   -   -  -   -
```

**Comment by angelrodriguez2020** [comment]
Dear Stef,

Thank you for your help.

Your 'post' solution works for numeric variables to predict aid, but not for factors. Please see what happens when I introduce a factor as a predictor in your code:

````r
data <- data.frame(
  age = c(60, 54, 82, NA, 34, 58, 60, 58, 76, 40),
  fall = c(0, NA, 1, 0, 1, NA, 1, 1, 0, 1),
  aid = factor(c("-", NA, "Yes", "-", "No", NA, "Yes", NA, NA, "No")),
  sex = factor(c("M", NA, "F", "F", "M", NA, "M", "F", "F", "M"))
)

pred <- make.predictorMatrix(data)
pred["fall", "aid"] <- 0
pred["age", "aid"] <- 0
pred["sex", "aid"] <- 0

post <- make.post(data)
post["aid"] <- paste(sep = "; ",
                     "fall <- data$fall[!r[,j]] == 1",
                     "idx <- data$fall == 1",
                     "imp[[j]][!fall, i] <- levels(data$aid)[1]",
                     "imp[[j]][fall, i] <- mice.impute.pmm(y = data$aid[idx], ry = r[idx, j], 
                     x = model.matrix(aid ~ 0 + age + sex , data[idx, , drop = FALSE]))")
                     
imp <- mice(data, m = 10, pred = pred, post = post, seed = 1)`

>imp <- mice(data, m = 10, pred = pred, post = post, seed = 1)

 iter imp variable
  1   1  age  fall  aid
 Error in get("printFlag", parent.frame(search.parents("printFlag"))) : 
  object 'printFlag' not found 
```

**Comment by stefvanbuuren** [comment]
The devil is in the details. Nothing is wrong with your code. 

Apparently, the introduction of `mice.impute.pmm()` at this - somewhat unusual - point in the algorithm surprised the function that finds the printing flag. We'll correct that.

**Comment by stefvanbuuren** [comment]
Solved in `mice 3.11.2`

**Comment by angelrodriguez2020** [comment]
Dear Stef,
Thank you.
I can update to mice 3.11.0, not to mice 3.11.2. 
When would that be possible?
Ángel

**Comment by stefvanbuuren** [comment]
See <https://github.com/amices/mice>

**Comment by angelrodriguez2020** [comment]
Thank you Stef, but I couldn´t. This is what I got (R 4.0.2, Windows 10):

> devtools::install_github(repo = "amices/mice")
Downloading GitHub repo amices/mice@HEAD
These packages have more recent versions available.
It is recommended to update all of them.
Which would you like to update?

1: All                              
2: CRAN packages only               
3: None                             
4: vctrs     (0.3.3 -> 0.3.4) [CRAN]
5: backports (1.1.7 -> 1.1.9) [CRAN]

Enter one or more numbers, or an empty line to skip updates:1
vctrs     (0.3.3 -> 0.3.4) [CRAN]
backports (1.1.7 -> 1.1.9) [CRAN]
Installing 2 packages: vctrs, backports
Installing packages into ‘C:/Users/arodr/Documents/R/win-library/4.0’
(as ‘lib’ is unspecified)

  There are binary versions available but the source versions are later:
          binary source needs_compilation
vctrs      0.3.3  0.3.4              TRUE
backports  1.1.7  1.1.9              TRUE

installing the source packages ‘vctrs’, ‘backports’

trying URL 'https://ftp.cixug.es/CRAN/src/contrib/vctrs_0.3.4.tar.gz'
Content type 'application/x-gzip' length 995078 bytes (971 KB)
downloaded 971 KB

trying URL 'https://ftp.cixug.es/CRAN/src/contrib/backports_1.1.9.tar.gz'
Content type 'application/x-gzip' length 18903 bytes (18 KB)
downloaded 18 KB

* installing *source* package 'vctrs' ...
** package 'vctrs' successfully unpacked and MD5 sums checked
ERROR: cannot remove earlier installation, is it in use?
* removing 'C:/Users/arodr/Documents/R/win-library/4.0/vctrs'
* restoring previous 'C:/Users/arodr/Documents/R/win-library/4.0/vctrs'
Error in file.copy(lp, dirname(pkgdir), recursive = TRUE, copy.date = TRUE) : 
  (converted from warning) problem copying C:\Users\arodr\Documents\R\win-library\4.0\00LOCK-vctrs\vctrs\libs\x64\vctrs.dll to C:\Users\arodr\Documents\R\win-library\4.0\vctrs\libs\x64\vctrs.dll: Permission denied
* removing 'C:/Users/arodr/Documents/R/win-library/4.0/vctrs'
* restoring previous 'C:/Users/arodr/Documents/R/win-library/4.0/vctrs'
Error in file.copy(lp, dirname(pkgdir), recursive = TRUE, copy.date = TRUE) : 
  (converted from warning) problem copying C:\Users\arodr\Documents\R\win-library\4.0\00LOCK-vctrs\vctrs\libs\x64\vctrs.dll to C:\Users\arodr\Documents\R\win-library\4.0\vctrs\libs\x64\vctrs.dll: Permission denied
Execution halted
Error: Failed to install 'mice' from GitHub:
  (converted from warning) installation of package ‘vctrs’ had non-zero exit status
>

**Comment by stefvanbuuren** [comment]
Not related to `mice`, so please look elsewhere to solve this.

**Comment by LaurenceGeebelen** [comment]
Dear all,

I'm trying to solve a similar problem, however in my example aid would have an additional level, leaving 3 options when fall == 1  and I would like to use "polyreg" as a method for the imputation (as this is used for other multicategorical variables in my imputation model), I keep getting errors when adapting following code to polyreg , can someone help?  

post["aid"] <- paste(sep = "; ",
                     "fall <- data$fall[!r[,j]] == 1",
                     "idx <- data$fall == 1",
                     "imp[[j]][!fall, i] <- levels(data$aid)[1]",
                     "imp[[j]][fall, i] <- mice.impute.**pmm**(**y = data$aid[idx], ry = r[idx, j], x = model.matrix(aid ~ 0 + age, data[idx, , drop = FALSE]))**")

Thank you!
Laurence

**Reply by alan-y** [reply]
Hi @stefvanbuuren, thanks for your help with this problem (which I'm currently trying to solve as well for a piece of work). Could I ask what the `r` is referring to in this code please? I think I've worked out everything else in my head - I think `j` is referring to the `aid` column and  `i` is to the cells where falls do or don't happen?

```r
post["aid"] <- paste(sep = "; ",
                     "fall <- data$fall[!r[,j]] == 1",
                     "idx <- data$fall == 1",
                     "imp[[j]][!fall, i] <- levels(data$aid)[1]",
                     "imp[[j]][fall, i] <- mice.impute.pmm(y = data$aid[idx], ry = r[idx, j], x = model.matrix(aid ~ 0 + age, data[idx, , drop = FALSE]))")
```

**Comment by stefvanbuuren** [comment]
See <https://github.com/amices/mice/blob/master/R/sampler.R> for the meaning of `i` and `j`

**Comment by HarryVrn** [comment]
I am currently looking for a solution for a similar problem. 

My data contains a variable paid_work with levels "Yes" and "No" and a continuous variable days_absent_work which is structurally missing for paid_job == "No". I want to impute days_absent_work for observations with paid_work == "Yes". There are 7 observations that did not answer days_absent_work with a paid job; however, there are 753 observations for which paid_job is missing and for which "Yes" may be imputed, thus also needing imputation for days_absent_work. 

I followed your examples and ran tryouts of code in AI in a process of trial and error, but never got a working solution. 

```
ini <- mice(final_auxil, maxit = 0, print = F)
pred <- ini$predictorMatrix 
pred[, c("id", "weight", "hh_cluster")] <- 0
pred[c("id", "weight", "hh_cluster"), ] <- 0
pred["days_absent_work", "paid_job"] <- 0
pred

post <- ini$post
post["days_absent_work"] <- paste(sep = "; ",
                                  "mis <- !r[, j]",
                                  "to_imp <- data$paid_job[mis] == 'Yes'",
                                  "donor <- data$paid_job == 'Yes' & r[, j]",
                                  "imp[[j]][!to_imp, i] <- 0",
                                  "imp[[j]][to_imp, i] <- mice.impute.pmm(y = data$days_absent_work[donor], ry = r[donor, j], x = model.matrix(days_absent_work ~ 0 + age + sex + regio + major_cov + total_cost + civil_status + nationality + urbanisation + hh_composition + number_hh + highest_educ, data[donor, , drop = FALSE]))"
                                  )

imp <- mice(final_auxil, m = 1, maxit = 20, post = post, predictorMatrix = pred, print = TRUE)
```

The error I received: 
`Error in x[[jj]][iseq] <- vjj : replacement has length zero`

Multiple times, I also received the error that `Error in lm.fit(...): 0 (non-NA) cases`
```
post["days_absent_work"] <- paste(sep = "; ",
"paid_job <- data$paid_job[!r[,j]] == 1",
"idx <- data$paid_job == 1",
"imp[[j]][!paid_job, i] <- 0",
"imp[[j]][paid_job, i] <- mice.impute.pmm(y = data$days_absent_work[idx], ry = r[idx, j], x = model.matrix(days_absent_work ~ 0 + age + sex + regio + major_cov + total_cost + civil_status + nationality + urbanisation + hh_composition + number_hh + highest_educ, data[idx, , drop = FALSE]))
```

The difference with the previous code is that, here, I mistakenly used 'paid_job == 1' while it is a variable with levels "Yes" and "No".

@stefvanbuuren I hope you can help me figure this out and I appreciate your engagement in this community, your work has helped me a lot.

**Reply by stefvanbuuren** [reply]
Quite an involved setup. I am not able to replicate.

In order to make progress, it would help to create a reprex with a small dataset, e.g. 20 rows and 5 columns that has all the features of your imputation problem.

**Reply by HarryVrn** [reply]
Thank you for your reply. 

I managed to get sensible imputations with this set-up:

```
final_auxil <- final_auxil %>%
  mutate(days_absent_work = if_else(paid_job == "No", 0, days_absent_work))

## Specify predictor matrix
ini <- mice(final_auxil, maxit = 0, print = F)
pred <- ini$predictorMatrix
pred[, c("id", "weight", "hh_cluster")] <- 0
pred[c("id", "weight", "hh_cluster"), ] <- 0
pred["paid_job", "days_absent_work"] <- 0

## Add post-processing step 
post <- ini$post
post["days_absent_work"] <- "imp[[j]][data$paid_job[!r[, j] & where[, j]] == 'No', i] <- 0"

## Impute 
set.seed(7)
imp <- mice(final_auxil, m = 1, maxit = 40, where = where, post = post, predictorMatrix = pred, print = TRUE)
```

Now I added smoking variables of interest, namely smoking_status and the exposure variable pack_years. Again there are dependencies in the data: pack-years should be 0 for never smokers and non-zero for daily, former and occasional smokers. There are no observed values for pack-years in former smokers because of how the survey was conducted (composite variables used to calculate pack years were not asked in former smokers). I chose to drop the composite variables and only use the final variable pack-years in the imputation model to keep it simpler. 

```
final_auxil <- final_auxil %>%
  mutate(days_absent_work = if_else(paid_job == "No", 0, days_absent_work),
         pack_years = if_else(smoking_status == "Never", 0, pack_years))

## Specify predictor matrix
ini <- mice(final_auxil, maxit = 0, print = F)
pred <- ini$predictorMatrix
pred[, c("id", "weight", "hh_cluster")] <- 0
pred[c("id", "weight", "hh_cluster"), ] <- 0
pred["paid_job", "days_absent_work"] <- 0
pred["smoking_status", "pack_years"] <- 0

## Add post-processing step 
post <- ini$post
post["days_absent_work"] <- "imp[[j]][data$paid_job[!r[, j] & where[, j]] == 'No', i] <- 0"
post["pack_years"] <- "imp[[j]][data$smoking_status[!r[, j] & where[, j]] == 'Never', i] <- 0"

## Impute 
set.seed(7)
imp3 <- mice(final_auxil, m = 1, maxit = 20, where = where, post = post, predictorMatrix = pred, method = meth, print = TRUE)
```

 This imputation model functions: smoking_status$Former gets dropped as a predictor for pack-years because there are no observed values for this category. 

The imputations seem sensible and follow the observed distribution of pack years for daily smokers, and the imputed values for former smokers seem to resemble that of daily smokers which is desirable. However, for occasional smokers, there are a lot of zero imputations while there are no observed zeroes in pack years for occasional smokers. Somehow the model borrows donors from never smokers to impute values for occasional smokers, while it does not do this for daily and former smokers. How is this possible, since smoking status is included as predictor for pack years? 

<img width="1471" height="874" alt="image" src="https://github.com/user-attachments/assets/7864e60c-1810-4231-b9b1-75f45fb3ceb8" />

Including this post-processing rule 
```
post["pack_years"] <- paste(
  "mis <- !r[, j] & where[, j]",
  "imp[[j]][data$smoking_status[mis] == 'Never', i] <- 0",
  "idx_occ <- data$smoking_status[mis] == 'Occasional'",
  "imp[[j]][idx_occ, i] <- squeeze(imp[[j]][idx_occ, i], c(0.00904, 10.7))",
  sep = ";\n"
)
```
just puts all those zero imputed values for occasional smokers to 0.00904, which is not really a solution since the values still do not follow the observed distribution of pack years in occasional smokers.

Do you have any idea why the model borrows zeroes for imputation of pack-years from never smokers for occasional smokers, but not for daily or former smokers? And how to prevent this from happening in occasional smokers? 

Thank you for your comment.

**Reply by stefvanbuuren** [reply]
Glad that you were able to make progress. It is an interesting, but complicated, problem.

You are stretching the limits of the mice post-processing. I am afraid I can't be of much help without a well-curated small example of the problem that I can run myself to see what's going on. Some suggestions: 

- Paste these syntaxes into chatgpt, understand what each step is doing, and ask for assistance;
- Write your own dedicated mice.impute.xxx() function for pack-years;
- Impute everything, and make specific entries NA or 0 in the completed datasets after imputation;
- Create a small, runable example, and try to figure out the principles from scratch (recommended).

Hope this helps to get you to sensible next steps.

**Reply by HarryVrn** [reply]
Hi Stef,

See here the code I used to create a running example with a data structure similar to the original dataset:
```
# Set seed for reproducibility
set.seed(128)

# Number of observations
n <- 1000

# Initialize dataframe
df <- data.frame(
  age = numeric(n),
  sex = character(n),
  regio = character(n),
  major_cov = character(n),
  total_cost = numeric(n),
  smoking_status = character(n),
  pack_years = numeric(n),
  time_quit_smoking = character(n),
  stringsAsFactors = FALSE
)

# Age: 18-85, no missingness
df$age <- sample(18:85, n, replace = TRUE)

# Sex: Male/Female, no missingness
df$sex <- sample(c("Male", "Female"), n, replace = TRUE)

# Region: no missingness
df$regio <- sample(c("North", "South", "East", "West"), n, replace = TRUE)

# Major_cov: 6% missingness
df$major_cov <- sample(c("Yes", "No"), n, replace = TRUE)
missing_idx <- sample(1:n, round(n * 0.06))
df$major_cov[missing_idx] <- NA

# Total_cost: random between 0-50000
df$total_cost <- sample(0:50000, n, replace = TRUE)

# Smoking_status: 15% daily, 4% occasional, 23% former, 55% never, 3% missing
smoking_probs <- c(0.15, 0.04, 0.23, 0.55, 0.03)
smoking_levels <- c("Daily", "Occasional", "Former", "Never", NA)
df$smoking_status <- sample(smoking_levels, n, replace = TRUE, prob = smoking_probs)

# Pack_years based on smoking_status
for (i in 1:n) {
  if (is.na(df$smoking_status[i])) {
    df$pack_years[i] <- NA
  } else if (df$smoking_status[i] == "Daily") {
    # 6% missingness for daily smokers
    if (runif(1) < 0.15) {
      df$pack_years[i] <- NA
    } else {
      df$pack_years[i] <- round(runif(1, 2, 3), 1)
    }
  } else if (df$smoking_status[i] == "Occasional") {
    # 40% missingness for occasional smokers
    if (runif(1) < 0.50) {
      df$pack_years[i] <- NA
    } else {
      df$pack_years[i] <- round(runif(1, 1, 2), 1)
    }
  } else if (df$smoking_status[i] == "Former") {
    # 100% missingness for former smokers
    df$pack_years[i] <- NA
  } else if (df$smoking_status[i] == "Never") {
    df$pack_years[i] <- 0
  }
}

# Time_quit_smoking
quit_options <- c(
  "Not applicable",
  "Less than 1 month ago",
  "1 month ago or longer, but less than 6 months ago",
  "6 months ago or longer, but less than 1 year ago",
  "1 year ago or longer, but less than 2 years ago",
  "2 years ago or longer, but less than 10 years ago",
  "10 years ago or longer"
)

for (i in 1:n) {
  if (is.na(df$smoking_status[i])) {
    df$time_quit_smoking[i] <- NA
  } else if (df$smoking_status[i] %in% c("Daily", "Occasional")) {
    df$time_quit_smoking[i] <- "Current"
  } else if (df$smoking_status[i] == "Former") {
    # 4% missingness for former smokers
    if (runif(1) < 0.1) {
      df$time_quit_smoking[i] <- NA
    } else {
      # Random selection excluding "Current"
      df$time_quit_smoking[i] <- sample(quit_options[2:7], 1)
    }
  } else if (df$smoking_status[i] == "Never") {
    df$time_quit_smoking[i] <- NA
  }
}

# Convert character columns with NA to proper NA (not "NA" string)
df$smoking_status[df$smoking_status == "NA"] <- NA
df$time_quit_smoking[df$time_quit_smoking == ""] <- NA

# Convert variables to factors
df$sex <- as.factor(df$sex)
df$regio <- as.factor(df$regio)
df$smoking_status <- as.factor(df$smoking_status)
df$time_quit_smoking <- as.factor(df$time_quit_smoking)
df$major_cov <- as.factor(df$major_cov)

df$time_quit_smoking <- factor(
  df$time_quit_smoking,
  levels = quit_options,
  ordered = TRUE
)
```

Based on the generated dataframe, no relationship between demographic variables and smoking status is expected (no data was generated conditionally on smoking status, except for pack years). 

```
### MICE
library(mice)
## Specify predictor matrix
df1 <- df %>% select(-time_quit_smoking)
ini <- mice(df1, maxit = 0, print = F)
pred <- ini$predictorMatrix
pred["smoking_status", "pack_years"] <- 0
pred

## Specify method
meth <- ini$method

## Add post-processing step for conditional imputation of days_absent_work
post <- ini$post
post["pack_years"] <- "imp[[j]][data$smoking_status[!r[, j] & where[, j]] == 'Never', i] <- 0"

## (Specify where matrix)
where <- ini$where

## Impute 
set.seed(7)
imp <- mice(df1, m = 1, maxit = 20, where = where, post = post, method = meth, predictorMatrix = pred, print = TRUE)
imp$loggedEvents

summary(complete(imp, 1))
complete_1 <- complete(imp, 1)

stripplot(imp, pack_years ~ smoking_status)
```

Based on this code, one would expect the stripplot to look like the following:
- Never smokers: all pack year values = 0
- Daily smokers: pack year values between 2 and 3
- Occasional smokers: pack year values between 1 and 2
- Former smokers: pack year values between 1 and 3

However, this is the output of the stripplot function: 
<img width="643" height="488" alt="Rplot" src="https://github.com/user-attachments/assets/22812721-bd66-49a3-83a6-6c5268b059cd" />

I don't understand why for former smokers values are only drawn from daily and not occasional smokers. I played with the prevalence of the smoking categories as well as amount of missingness, but always replicated this result. 

Would you have any idea on why this happens? 

Best,
